full-moon/src/tokenizer.rs

```diff
@@ -650,260 +642,153 @@ impl<'a> fmt::Display for StringLiteralQuoteType {
-fn parse_digit_with_seperator(code: &str) -> IResult<&str, &str> {
-    digit1(code)
-}
-
-#[cfg(feature = "roblox")]
-fn parse_digit_with_seperator(code: &str) -> IResult<&str, &str> {
-    recognize(pair(
-        digit1,
-        opt(take_while1(|c: char| c.is_digit(10) || c == '_')),
-    ))(code)
-}
-
-fn parse_no_int_fractional_number(code: &str) -> IResult<&str, &str> {
-    recognize(pair(
-        opt(parse_digit_with_seperator),
-        pair(
-            pair(tag("."), parse_digit_with_seperator),
-            opt(pair(
-                pair(tag_no_case("e"), opt(alt((tag("-"), tag("+"))))),
-                parse_digit_with_seperator,
-            )),
-        ),
-    ))(code)
 }
 
-fn parse_basic_number(code: &str) -> IResult<&str, &str> {
-    recognize(pair(
-        parse_digit_with_seperator,
-        pair(
-            opt(pair(tag("."), opt(parse_digit_with_seperator))),
-            opt(pair(
-                pair(tag_no_case("e"), opt(alt((tag("-"), tag("+"))))),
-                parse_digit_with_seperator,
-            )),
-        ),
-    ))(code)
-}
-
-#[cfg(feature = "roblox")]
-fn parse_roblox_number(code: &str) -> IResult<&str, &str> {
-    recognize(pair(
-        tag_no_case("0b"),
-        take_while1(|x: char| x == '0' || x == '1' || x == '_'),
-    ))(code)
-}
-
-fn parse_number(code: &str) -> IResult<&str, &str> {
-    alt((
-        #[cfg(feature = "roblox")]
-        parse_roblox_number,
-        parse_hex_number,
-        parse_basic_number,
-        parse_no_int_fractional_number,
-    ))(code)
-}
-
-fn advance_number(code: &str) -> Advancement {
-    match parse_number(code) {
-        Ok((_, number)) => Ok(Some(TokenAdvancement {
-            advance: number.chars().count(),
-            token_type: TokenType::Number {
-                text: Cow::from(number),
-            },
-        })),
-        Err(_) => Ok(None),
+impl<'a> From<TokenizerErrorType> for RawToken<'a> {
+    fn from(error: TokenizerErrorType) -> RawToken<'a> {
+        Err(error)
     }
 }
 
-#[inline]
-fn parse_identifier(code: &str) -> IResult<&str, &str> {
-    recognize(pair(
-        alt((alpha1, tag("_"))),
-        many0(alt((alphanumeric1, tag("_")))),
-    ))(code)
-}
-
-fn advance_identifier(code: &str) -> Advancement {
-    match parse_identifier(code) {
-        Ok((_, identifier)) => Ok(Some(TokenAdvancement {
-            advance: identifier.chars().count(),
-            token_type: TokenType::Identifier {
-                identifier: Cow::from(identifier),
-            },
-        })),
-        Err(_) => Ok(None),
-    }
-}
+peg::parser! {
+    grammar tokens() for str {
+        use super::ParseSymbol;
+        use peg::ParseLiteral;
+        use super::StringLiteralQuoteType as QuoteType;
+        rule line_ending()
+            = "\n" / "\r\n"
+        rule space()
+            = [' '|'\t']
+
+        pub(super) rule whitespace() -> RawToken<'input>
+            = chars:$( space()+ line_ending()? / line_ending() )
```

Comment:
```suggestion
            = chars:$( space() + line_ending()? / line_ending() )
```

Reply:
This is a trailing operator for 1 or more repeats, not addition.

Reply:
Oops, good catch.